java-checks/src/main/java/org/sonar/java/checks/UseSwitchExpressionCheck.java

```diff
@@ -0,0 +1,127 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Collections;
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.java.JavaVersionAwareVisitor;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaVersion;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.CaseGroupTree;
+import org.sonar.plugins.java.api.tree.ExpressionStatementTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.SwitchStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import static org.sonar.plugins.java.api.tree.Tree.Kind.MEMBER_SELECT;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.RETURN_STATEMENT;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.THROW_STATEMENT;
+
+@Rule(key = "S5194")
+public class UseSwitchExpressionCheck extends IssuableSubscriptionVisitor implements JavaVersionAwareVisitor {
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.SWITCH_STATEMENT);
+  }
+
+  @Override
+  public boolean isCompatibleWithJavaVersion(JavaVersion version) {
+    return version.isJava12Compatible();
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    SwitchStatementTree switchStatementTree = (SwitchStatementTree) tree;
+    Symbol switchAssigningVariable = isSwitchAssigningVariable(switchStatementTree);
+    if (switchAssigningVariable != null) {
+      reportIssue(switchStatementTree.switchKeyword(), "Use expression switch to set value of '" + switchAssigningVariable.name() + "'.");
+    } else if (isSwitchReturningValue(switchStatementTree)) {
+      reportIssue(switchStatementTree.switchKeyword(), "Use expression switch to return the value.");
+    }
+  }
+
+  private static Symbol isSwitchAssigningVariable(SwitchStatementTree switchStatementTree) {
+    Symbol switchAssigningVariable = null;
+    for (CaseGroupTree caseGroupTree : switchStatementTree.cases()) {
+      if (hasSingleStatement(caseGroupTree.body(), THROW_STATEMENT)) {
+        continue;
+      }
+      Symbol variable = assigningVariable(caseGroupTree.body());
+      if (variable == null) {
+        return null;
+      }
+      if (switchAssigningVariable == null) {
+        switchAssigningVariable = variable;
+      } else if (switchAssigningVariable != variable) {
```

Comment:
what about unknown symbols? It seems to me that if symbols are always unknown, then we will have FPs:

```java
switch(expr) {
  case 1:
    unknown1 = 42;
    break;
  case 2:
    unkown2 = 666;
    break;
}
```